File: Scripts/Python/tile-with-clip.py
# Tile current selection with clipboard pattern.
# Author: Andrew Trevorrow (andrew@trevorrow.com), March 2006.
# Updated to use exit command, Nov 2006.
# Updated to handle multi-state patterns, Aug 2008.

from glife import *
import golly as g

# assume one-state cell list (may change below)
inc = 2

# ------------------------------------------------------------------------------

def clip_rb (patt, right, bottom):
    # remove any cells outside given right and bottom edges
    clist = list(patt)

    #  remove padding int if present
    if (inc == 3) and (len(clist) % 3 == 1): clist.pop()

    x = 0
    y = 1
    while x < len(clist):
        if (clist[x] > right) or (clist[y] > bottom):
            # remove cell from list
            clist[x : x+inc] = []
        else:
            x += inc
            y += inc

    # append padding int if necessary
    if (inc == 3) and (len(clist) & 1 == 0): clist.append(0)

    return pattern(clist)

# ------------------------------------------------------------------------------

selrect = rect( g.getselrect() )
if selrect.empty: g.exit("There is no selection.")

cliplist = g.getclip()                    # 1st 2 items are wd,ht
pbox = rect( [0, 0] + cliplist[0 : 2] )
cliplist[0 : 2] = []                      # remove wd,ht
p = pattern( cliplist )

if len(cliplist) & 1 == 1: inc = 3        # multi-state?

g.clear(inside)
if len(cliplist) > 0:
    # tile selrect with p, clipping right & bottom edges if necessary
    y = selrect.top
    while y <= selrect.bottom:
        bottom = y + pbox.height - 1
        x = selrect.left
        while x <= selrect.right:
            right = x + pbox.width - 1
            if (right <= selrect.right) and (bottom <= selrect.bottom):
                p.put(x, y)
            else:
                clip_rb( p(x, y), selrect.right, selrect.bottom ).put()
            x += pbox.width
        y += pbox.height

if not selrect.visible(): g.fitsel()

File: Scripts/Lua/tile-with-clip.lua
-- Tile current selection with clipboard pattern.
-- Author: Andrew Trevorrow (andrew@trevorrow.com), Mar 2016.

local g = golly()
local gp = require "gplus"

-- assume one-state cell array (may change below)
local multistate = false

--------------------------------------------------------------------------------

local function clip_rb(cells, right, bottom)
    -- set given cells except those outside given right and bottom edges
    local len = #cells
    local x = 1
    local y = 2
    if multistate then
        -- ignore padding int if present
        if len % 3 == 1 then len = len - 1 end
        while x <= len do
            if (cells[x] <= right) and (cells[y] <= bottom) then
                g.setcell(cells[x], cells[y], cells[x+2])
            end
            x = x + 3
            y = y + 3
        end
    else   
        while x <= len do
            if (cells[x] <= right) and (cells[y] <= bottom) then
                g.setcell(cells[x], cells[y], 1)
            end
            x = x + 2
            y = y + 2
        end
    end
end

--------------------------------------------------------------------------------

local selrect = g.getselrect()
if #selrect == 0 then g.exit("There is no selection.") end

-- get selection edges
local selleft, seltop, selright, selbottom = gp.getedges(selrect)

local pwidth, pheight, p = g.getclip()

if #p & 1 == 1 then multistate = true end

g.clear(0)
if #p > 0 then
    -- tile selrect with p, clipping right & bottom edges if necessary
    local y = seltop
    while y <= selbottom do
        local bottom = y + pheight - 1
        local x = selleft
        while x <= selright do
            local right = x + pwidth - 1
            if (right <= selright) and (bottom <= selbottom) then
                g.putcells(p, x, y)
            else
                local tempcells = g.transform(p, x, y)
                clip_rb(tempcells, selright, selbottom)
            end
            x = x + pwidth
        end
        y = y + pheight
    end
end
if not g.visrect(selrect) then g.fitsel() end
